src/proxy/resolve.rs

```diff
@@ -107,61 +125,442 @@ where
 
     fn call(&mut self, target: T) -> Self::Future {
         let resolution = self.resolve.resolve(&target);
-        futures::future::ok(Discover {
+        futures::future::ok(Discover::new(resolution, self.inner.clone()))
+    }
+}
+
+// === impl Discover ===
+
+impl<R, M> Discover<R, M>
+where
+    R: Resolution,
+    M: svc::Service<R::Endpoint>,
+{
+    fn new(resolution: R, make: M) -> Self {
+        Self {
             resolution,
-            make: self.inner.clone(),
+            make,
+            makes: MakeStream::new(),
             is_empty: Arc::new(AtomicBool::new(false)),
-        })
+        }
     }
 }
 
 impl<R, M> HasEndpointStatus for Discover<R, M>
 where
     R: Resolution,
+    M: svc::Service<R::Endpoint>,
 {
     fn endpoint_status(&self) -> EndpointStatus {
         EndpointStatus(self.is_empty.clone())
     }
 }
 
-impl<R, M> tower_discover::Discover for Discover<R, M>
+impl<R, M> Discover<R, M>
 where
     R: Resolution,
     R::Endpoint: fmt::Debug,
     R::Error: Into<Error>,
-    M: rt::Make<R::Endpoint>,
+    M: svc::Service<R::Endpoint>,
+    M::Error: Into<Error>,
 {
-    type Key = SocketAddr;
-    type Service = M::Value;
-    type Error = Error;
-
-    fn poll(&mut self) -> Poll<Change<Self::Key, Self::Service>, Self::Error> {
+    fn poll_resolution(&mut self) -> Poll<Change<SocketAddr, M::Response>, Error> {
         loop {
+            try_ready!(self.make.poll_ready().map_err(Into::into));
```

Comment:
We need to check for `make`'s readiness, because this is the readiness of the Discover for the endpoint stack? Once we know that is readiness, we can poll for updates?

Reply:
I've added a comment:
```
            // Before polling the resolution, where we could potentially receive
            // an `Add`, poll_ready to ensure that `make` is ready to build new
            // services. Don't process any updates until we can do so.
```